### PR TITLE
[QLever] Add qlever_jll binary

### PR DESCRIPTION
**Pull request by jeremiahpslewis**

### CODE REVIEW

File: Q/qlever/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+# The products that we will ensure are always built
+products = Product[
+    LibraryProduct(["libjemalloc", "jemalloc"], :libjemalloc)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Libuuid_jll", uuid="38a345b3-de98-5d2b-a5d3-14cd9215e700"))
+    Dependency(PackageSpec(name="Git_jll", uuid="f8c6e375-362e-5223-8a59-34ff63f689eb"))
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:43]: Git is a dependency? 🤔

File: Q/qlever/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    Dependency(PackageSpec(name="Zstd_jll", uuid="3161d3a3-bdf6-5164-811a-617609db77b4"))
+    Dependency(PackageSpec(name="boost_jll", uuid="28df3c45-c428-5900-9ff8-a3135698ca75"))
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:45]: Must add compat

File: Q/qlever/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    Dependency(PackageSpec(name="ICU_jll", uuid="a51ab1cf-af8e-5615-a023-bc2c838bba6b"))
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:46]: Must add compat

File: Q/qlever/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    Dependency(PackageSpec(name="jemalloc_jll", uuid="454a8cc1-5e0e-5123-92d5-09b094f0e876"))
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version = v"11.1.0")
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:51]: Gcc 11 is likely to be incompatible with most systems

File: Q/qlever/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+make -j $(nproc)
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms(; experimental=true)
+
+# The products that we will ensure are always built
+products = Product[
+    LibraryProduct(["libjemalloc", "jemalloc"], :libjemalloc)
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:37]: ?

**Review by giordano — CHANGES_REQUESTED** [review]

File: Q/qlever/build_tarballs.jl (hunk 4 above)
Comment (giordano) [reply Q/qlever/build_tarballs.jl:45]: Sigh. Well, if the library ends up having symbols versioned with `GLIBCXX_3.4.29` then good luck dlopening it with julia v1.6

File: Q/qlever/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+script = raw"""
+cd $WORKSPACE/srcdir/qlever/
+
+git config --global advice.detachedHead false
+git submodule update --init --recursive
+
+mkdir build && cd build
+cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DUSE_PARALLEL=true -DABSL_PROPAGATE_CXX_STD=ON -DSTXXL_TRY_COMPILE_HEADERS=ON ..
+
+make -j $(nproc)
```
Comment (giordano) [inline Q/qlever/build_tarballs.jl:24]: I'm away from computer, but I seem to remember there is no `nproc` command
```suggestion
make -j ${nproc}
```

Comment (jeremiahpslewis) [reply Q/qlever/build_tarballs.jl]: @giordano Thanks!I copied this from the Yggdrasil codebase...is it worth/necessary to change the other instances as well?

- https://github.com/JuliaPackaging/Yggdrasil/blob/139747aa43faba7cc259e365a364d26af88c37e5/C/COSMA/build_tarballs.jl#L62
- https://github.com/JuliaPackaging/Yggdrasil/blob/139747aa43faba7cc259e365a364d26af88c37e5/F/faust/build_tarballs.jl#L70
- https://github.com/JuliaPackaging/Yggdrasil/blob/139747aa43faba7cc259e365a364d26af88c37e5/U/UBPF/build_tarballs.jl#L22
- https://github.com/jeremiahpslewis/Yggdrasil/blob/6c6f78cede7d3ca943dc58b937d1c001fb1343fa/U/Unicorn/build_tarballs.jl#L17
- https://github.com/jeremiahpslewis/Yggdrasil/blob/97ac9e8759c5d54355ae54d830a8a80628f5a132/X/XPA/build_tarballs.jl#L20

File: Q/QLever/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+# platforms are passed in on the command line
+platforms = supported_platforms(; experimental=true)
+platforms = expand_cxxstring_abis(platforms)
+
+# QLever depends on FOXXLL which only builds on 64-bit systems
+# https://github.com/stxxl/foxxll/blob/a4a8aeee64743f845c5851e8b089965ea1c219d7/foxxll/common/types.hpp#L25
+filter!(p -> nbits(p) != 32, platforms)
+
+# Building against musl on Linux blocked by tlx dependency, issue #36 (https://github.com/tlx/tlx/issues/36)
+ filter!(p -> !(Sys.islinux(p) && libc(p) == "musl"), platforms)
```
Comment (giordano) [inline Q/QLever/build_tarballs.jl:88]: Indentation is off
```suggestion
# Building against musl on Linux blocked by tlx dependency (https://github.com/tlx/tlx/issues/36)
filter!(p -> !(Sys.islinux(p) && libc(p) == "musl"), platforms)
```

File: Q/QLever/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+# The products that we will ensure are always built
+products = Product[
```
Comment (giordano) [inline Q/QLever/build_tarballs.jl:91]:
```suggestion
products = [
```

**Comment by jeremiahpslewis** [comment]
@giordano Thanks for the review! I'm running up against a wall with CMake...there's an issue with pthread that I cannot seem to debug, any thoughts?

```bash
sandbox:${WORKSPACE}/srcdir/qlever/build # cat ${WORKSPACE}/srcdir/qlever/build/CMakeFiles/CMakeError.log
Performing C SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: /workspace/srcdir/qlever/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_274eb && [1/2] Building C object CMakeFiles/cmTC_274eb.dir/src.c.o
[2/2] Linking C executable cmTC_274eb
ninja: job failed: : && /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11/x86_64-linux-gnu-gcc --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/ -DCMAKE_HAVE_LIBC_PTHREAD   CMakeFiles/cmTC_274eb.dir/src.c.o  -o cmTC_274eb   && :
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: CMakeFiles/cmTC_274eb.dir/src.c.o: in function `main':
src.c:(.text+0x2d): undefined reference to `pthread_create'
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: src.c:(.text+0x39): undefined reference to `pthread_detach'
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: src.c:(.text+0x45): undefined reference to `pthread_cancel'
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: src.c:(.text+0x56): undefined reference to `pthread_join'
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/11.1.0/../../../../x86_64-linux-gnu/bin/ld: src.c:(.text+0x6a): undefined reference to `pthread_atfork'
collect2: error: ld returned 1 exit status
ninja: subcommand failed


Source file was:
#include <pthread.h>

void* test_func(void* data)
{
  return data;
}

int main(void)
{
  pthread_t thread;
  pthread_create(&thread, NULL, test_func, NULL);
  pthread_detach(thread);
  pthread_cancel(thread);
  pthread_join(thread, NULL);
  pthread_atfork(NULL, NULL, NULL);
  pthread_exit(NULL);

  return 0;
}
```

What is failing is a CMake threads check, ([source here](https://github.com/Kitware/CMake/blob/2f37c48cf16d3afebddf6743b6ff55391a4cac83/Modules/FindThreads.cmake#L78:L90)), but it's not clear to me why.

**Comment by giordano** [comment]
To compile that file you need either `-pthread` or `-lpthread`

**Comment by jeremiahpslewis** [comment]
> To compile that file you need either -pthread or -lpthread (preferably -pthread)

Thanks! Where does it get specified exactly?

**Comment by giordano** [comment]
In the compiler invocation:
```
/opt/bin/x86_64-linux-gnu-libgfortran5-cxx11/x86_64-linux-gnu-gcc --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/ -DCMAKE_HAVE_LIBC_PTHREAD   CMakeFiles/cmTC_274eb.dir/src.c.o  -o cmTC_274eb
```
should be
```
/opt/bin/x86_64-linux-gnu-libgfortran5-cxx11/x86_64-linux-gnu-gcc --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/ -DCMAKE_HAVE_LIBC_PTHREAD   CMakeFiles/cmTC_274eb.dir/src.c.o  -o cmTC_274eb -pthread
```

>  E.g. this PR in the project seems to have removed most instances of -pthread: ad-freiburg/qlever#82

Maybe because you're supposed to set `-DCMAKE_THREAD_LIBS_INIT=-pthread`?

**Comment by jeremiahpslewis** [comment]
Seems like this issue arises for us (and not for the upstream project) because Yggdrasil cross-compiles. Based on this hunch, trying the following:

```bash
CMAKE_FLAGS+=(-DCMAKE_CXX_COMPILER="/opt/${target}/bin/${target}-g++  -pthread")
```

https://stackoverflow.com/questions/21892427/how-to-get-pthread-handling-right

**Comment by giordano** [comment]
That doesn't sound right, since we set `CMAKE_CXX_COMPILER` already in the toolchain file.  Maybe you should investigate why `CMAKE_THREAD_LIBS_INIT` is not being set to the right value?

**Comment by jeremiahpslewis** [comment]
@joka921 Updated the commit -> let’s see if it uses clang for the builds & what passes! :)

**Comment by giordano** [comment]
What are you doing with the `.ci` project?

**Comment by jeremiahpslewis** [comment]
> What are you doing with the `.ci` project?

Git caught it, didn't mean to commit. Semi-related, BinaryBuilder builds only work for me if I run them with `--project=.ci` specified.

**Pull request closed without merging**